source/request-as-event-emitter.js

```diff
@@ -150,7 +157,9 @@ module.exports = (options = {}) => {
 
 				const socket = req.connection;
 				if (socket) {
-					const onSocketConnect = () => {
+					const onSocketConnect = async () => {
```

Comment:
I'm not sure what the origin of this conditional block are, but I think the right way to handle this event is to remove the if block and make this:

```js
req.on('socket', () => { ... });
```

https://nodejs.org/dist/latest-v8.x/docs/api/http.html#http_event_socket

Reply:
Actually, since `onSocketConnect` and `onAbort` should be attached to the request, I don't think these are sensible hooks for `got` to configure directly. If you want to configure these for a got instance, you really just need to configure a listener for `request` and attach the request event listeners in the request listener.

Reply:
> I'm not sure what the origin of this conditional block are, but I think the right way to handle this event is to remove the if block

#429 

> Actually, since onSocketConnect and onAbort should be attached to the request, I don't think these are sensible hooks for got to configure directly. If you want to configure these for a got instance, you really just need to configure a listener for request and attach the request event listeners in the request listener.

Why would you do the same thing twice?

![wise](https://user-images.githubusercontent.com/36894700/42771060-407c25e8-8926-11e8-9497-f495ce14fc8f.png)

Reply:
Thanks for the reference to #429. Maybe I'm missing something, but I don't think that was the right fix b/c `if req.connection` is just testing a race condition. Shouldn't it be:

```js
req.once('socket', (sock) => {
  sock.once('connect', conn => {
    // ...
  }
};
```

I can see the convenience of low-friction hooks. Maybe there's a way to make the configuration a sort of DSL that got can use to wire these up. Something like this:

```js
{
  events: {
    request: {
      ['on|once']: {
        socket: {
          null: [(sock) => { console.log('request.on(socket)') }],
          connect: [(conn) => { console.log('socket.on(connect)') }]
        }
      },
    }
  }
}
```

With the above, the configuration can be traversed and the listeners can be subscribed without having to add handling code to got for every event.

I used `null` as a sentinel value to mean the subscriber for the event as opposed to a child subscription. Something else (symbol?) would work also.

Reply:
> Shouldn't it be:

Reply:
> I can see the convenience of low-friction hooks.

Great! We'll discuss this later. I was gonna push a PR with rewrited hooks from scratch but I decided current implementation is better. :)